File: pjecalc-fonte/negocio/br/jus/trt8/pjecalc/negocio/dominio/calculo/TabelaDeJurosDoCalculo.java
/*
 * Decompiled with CFR 0.152.
 */
package br.jus.trt8.pjecalc.negocio.dominio.calculo;

import br.jus.trt8.pjecalc.base.comum.HelperDate;
import br.jus.trt8.pjecalc.base.comum.Periodo;
import br.jus.trt8.pjecalc.base.comum.Utils;
import br.jus.trt8.pjecalc.negocio.comum.PeriodoDeJuros;
import br.jus.trt8.pjecalc.negocio.comum.TabelaDeJuros;
import br.jus.trt8.pjecalc.negocio.constantes.JurosDoAjuizamentoEnum;
import br.jus.trt8.pjecalc.negocio.dominio.calculo.Calculo;
import java.math.BigDecimal;
import java.util.Date;

public class TabelaDeJurosDoCalculo
extends TabelaDeJuros {
    private static final long serialVersionUID = 1L;
    private Date dataInicialDeJuros;
    private JurosDoAjuizamentoEnum jurosDoAjuizamento;
    private Periodo periodoDeJurosPadrao;
    private Periodo periodoDeJurosFazendaPublica;
    private boolean periodoDeJurosCalculado = false;

    public TabelaDeJurosDoCalculo(Calculo calculo) {
        super(calculo);
    }

    public TabelaDeJurosDoCalculo(Calculo calculo, Date dataInicialParaCalculo, Date dataFinalParaCalculo) {
        super(calculo, dataInicialParaCalculo, dataFinalParaCalculo);
    }

    @Override
    protected JurosDoAjuizamentoEnum getTipoDeJurosDoAjuizamento() {
        if (Utils.nulo((Object)this.jurosDoAjuizamento)) {
            return super.getTipoDeJurosDoAjuizamento();
        }
        return this.jurosDoAjuizamento;
    }

    @Override
    protected Date calcularDataInicialDoPrimeiroPeriodoDeJuros(Date dataInicialParaCalculo, Date dataFimVencimentoOcorrencia) {
        if (JurosDoAjuizamentoEnum.OCORRENCIAS_VENCIDAS.equals((Object)this.getTipoDeJurosDoAjuizamento())) {
            HelperDate dataInicial = HelperDate.getInstance(dataInicialParaCalculo).lastDayOfTheMonth();
            if (this.getCalculo().getParametrosDeAtualizacao().getAplicarJurosFasePreJudicial().booleanValue()) {
                Date dataAux = dataFimVencimentoOcorrencia;
                if (Utils.nulo(dataAux)) {
                    dataAux = dataInicial.getDate();
                }
                this.dataInicialDeJuros = dataAux;
                return HelperDate.getInstance(dataAux).addDay(1).getDate();
            }
            if (Utils.nulo(this.getCalculo().getDataAjuizamento()) || dataInicial.greaterThenOrEquals(this.getCalculo().getDataAjuizamento())) {
                if (Utils.naoNulo(dataFimVencimentoOcorrencia) && HelperDate.dateBeforeOrEquals(dataFimVencimentoOcorrencia, this.getCalculo().getDataAjuizamento())) {
                    this.dataInicialDeJuros = this.getCalculo().getDataAjuizamento();
                    return HelperDate.getInstance(this.getCalculo().getDataAjuizamento()).addDay(1).getDate();
                }
                if (Utils.naoNulo(dataFimVencimentoOcorrencia)) {
                    this.dataInicialDeJuros = dataFimVencimentoOcorrencia;
                    return HelperDate.getInstance(dataFimVencimentoOcorrencia).addDay(1).getDate();
                }
                this.dataInicialDeJuros = dataInicial.getDate();
                return dataInicial.addDay(1).getDate();
            }
            this.dataInicialDeJuros = this.getCalculo().getDataAjuizamento();
            return HelperDate.getInstance(this.getCalculo().getDataAjuizamento()).addDay(1).getDate();
        }
        this.dataInicialDeJuros = this.getCalculo().getDataAjuizamento();
        return HelperDate.getInstance(this.getCalculo().getDataAjuizamento()).addDay(1).getDate();
    }

    @Override
    public Date getDataInicialDeJuros() {
        return this.dataInicialDeJuros;
    }

    @Override
    public void setDataInicialDeJuros(Date dataInicialDeJuros) {
        this.dataInicialDeJuros = dataInicialDeJuros;
    }

    private TabelaDeJurosDoCalculo definirPeriodosDeJuros() {
        if (!this.periodoDeJurosCalculado) {
            PeriodoDeJuros periodo = super.getPeriodoDeJurosInicial();
            if (Utils.naoNulo(periodo)) {
                if (!periodo.isFazendaPublica()) {
                    this.periodoDeJurosPadrao = new Periodo(periodo.getDataInicial(), null);
                    while (Utils.naoNulo(periodo) && !periodo.isFazendaPublica()) {
                        this.periodoDeJurosPadrao.setFinal(periodo.getDataFinal());
                        periodo = periodo.getProximoPeriodo();
                    }
                }
                if (Utils.naoNulo(periodo) && periodo.isFazendaPublica()) {
                    this.periodoDeJurosFazendaPublica = new Periodo(periodo.getDataInicial(), null);
                    while (Utils.naoNulo(periodo) && periodo.isFazendaPublica()) {
                        this.periodoDeJurosFazendaPublica.setFinal(periodo.getDataFinal());
                        periodo = periodo.getProximoPeriodo();
                    }
                }
            }
            this.periodoDeJurosCalculado = true;
        }
        return this;
    }

    public Periodo getPeriodoDeJurosPadrao() {
        return this.definirPeriodosDeJuros().periodoDeJurosPadrao;
    }

    public Periodo getPeriodoDeJurosFazendaPublica() {
        return this.definirPeriodosDeJuros().periodoDeJurosFazendaPublica;
    }

    public BigDecimal calcularTaxaDeJuros(Date data, Date dataFimVencimentoOcorrencia, JurosDoAjuizamentoEnum jurosDoAjuizamento, boolean projetarData, boolean isMultaHonorario) {
        this.jurosDoAjuizamento = jurosDoAjuizamento;
        return super.calcularTaxaDeJuros(data, dataFimVencimentoOcorrencia, projetarData, isMultaHonorario);
    }

    public BigDecimal calcularTaxaDeJurosPagamento(Date data, Date dataFinal, JurosDoAjuizamentoEnum jurosDoAjuizamento, boolean projetarData, boolean isMultaHonorario) {
        this.jurosDoAjuizamento = jurosDoAjuizamento;
        return super.calcularTaxaDeJurosPagamento(data, dataFinal, projetarData, isMultaHonorario);
    }
}

File: src/lib/pjecalc/core/dominio/calculo/tabela-de-juros-do-calculo.ts
/**
 * PJe-Calc v2.15.1 — TabelaDeJurosDoCalculo (stub estrutural)
 * Porte de: br.jus.trt8.pjecalc.negocio.dominio.calculo.TabelaDeJurosDoCalculo
 *
 * Ref Java: pjecalc-fonte/.../calculo/TabelaDeJurosDoCalculo.java (~124 linhas)
 *
 * Extende `TabelaDeJuros` (fase 5) com informações específicas do Calculo:
 *   - jurosDoAjuizamento (OCORRENCIAS_VENCIDAS / TODAS)
 *   - dataInicialDeJuros
 *   - periodoDeJurosPadrao / periodoDeJurosFazendaPublica (calculado lazy)
 *
 * **Status**: stub — a lógica de `calcularDataInicialDoPrimeiroPeriodoDeJuros`
 * e `definirPeriodosDeJuros` depende do `PeriodoDeJuros.proximoPeriodo` e das
 * flags do ParametrosDeAtualizacao. Implementação virá quando MaquinaDeCalculo
 * precisar de juros calculados.
 */
import Decimal from 'decimal.js';
import { Periodo } from '../../base/comum/periodo';
import { JurosDoAjuizamentoEnum } from '../../constantes/enums';
import type { PeriodoDeJuros } from '../../comum/periodo-de-juros';
import { TABELA_SELIC_MENSAL } from '../indices/selic/tabela-selic-mensal';
import type { Calculo } from './calculo';

export class TabelaDeJurosDoCalculo {
  private calculo: Calculo;
  private dataInicialParaCalculo: Date | null = null;
  private dataFinalParaCalculo: Date | null = null;
  private dataInicialDeJuros: Date | null = null;
  private jurosDoAjuizamento: JurosDoAjuizamentoEnum | null = null;
  private periodoDeJurosPadrao: Periodo | null = null;
  private periodoDeJurosFazendaPublica: Periodo | null = null;
  private periodoDeJurosCalculado: boolean = false;

  constructor(
    calculo: Calculo,
    dataInicialParaCalculo: Date | null = null,
    dataFinalParaCalculo: Date | null = null,
  ) {
    this.calculo = calculo;
    this.dataInicialParaCalculo = dataInicialParaCalculo;
    this.dataFinalParaCalculo = dataFinalParaCalculo;
  }

  getCalculo(): Calculo { return this.calculo; }

  getDataInicialParaCalculo(): Date | null { return this.dataInicialParaCalculo; }
  setDataInicialParaCalculo(d: Date | null): void { this.dataInicialParaCalculo = d; }

  getDataFinalParaCalculo(): Date | null { return this.dataFinalParaCalculo; }
  setDataFinalParaCalculo(d: Date | null): void { this.dataFinalParaCalculo = d; }

  getDataInicialDeJuros(): Date | null { return this.dataInicialDeJuros; }
  setDataInicialDeJuros(d: Date | null): void { this.dataInicialDeJuros = d; }

  getJurosDoAjuizamento(): JurosDoAjuizamentoEnum | null { return this.jurosDoAjuizamento; }
  setJurosDoAjuizamento(v: JurosDoAjuizamentoEnum | null): void { this.jurosDoAjuizamento = v; }

  getPeriodoDeJurosPadrao(): Periodo | null {
    return this.periodoDeJurosPadrao;
  }

  getPeriodoDeJurosFazendaPublica(): Periodo | null {
    return this.periodoDeJurosFazendaPublica;
  }

  /**
   * definirPeriodosDeJuros — porte 1:1 de
   * `TabelaDeJurosDoCalculo.definirPeriodosDeJuros` (Java linhas 82-104).
   *
   * Percorre a cadeia de PeriodoDeJuros (raiz → proximoPeriodo*) e separa em:
   *   - `periodoDeJurosPadrao`: bloco contíguo inicial onde isFazendaPublica()=false.
   *     Período resultante: [primeiro.dataInicial, último_padrao.dataFinal].
   *   - `periodoDeJurosFazendaPublica`: bloco contíguo subsequente onde
   *     isFazendaPublica()=true. Período: [primeiro_fp.dataInicial, último_fp.dataFinal].
   *
   * Se o primeiro período já é fazenda pública, `periodoDeJurosPadrao` permanece
   * null (paridade Java linha 86 `if (!periodo.isFazendaPublica())`).
   *
   * Idempotente — só recalcula uma vez (`periodoDeJurosCalculado`).
   *
   * @param raiz raiz da cadeia de PeriodoDeJuros (vem do parent TabelaDeJuros).
   */
  definirPeriodosDeJuros(raiz: PeriodoDeJuros | null): this {
    if (this.periodoDeJurosCalculado) return this;
    let periodo: PeriodoDeJuros | null = raiz;
    if (periodo !== null) {
      // Bloco padrão (não fazenda pública).
      if (!periodo.isFazendaPublica()) {
        const inicial = periodo.getDataInicial();
        if (inicial !== null) {
          this.periodoDeJurosPadrao = new Periodo(inicial, null);
          while (periodo !== null && !periodo.isFazendaPublica()) {
            const fim: Date | null = periodo.getDataFinal();
            if (fim !== null) this.periodoDeJurosPadrao.setFinal(fim);
            periodo = periodo.getProximoPeriodo();
          }
        }
      }
      // Bloco fazenda pública (após o padrão, ou direto se a cadeia começa em FP).
      if (periodo !== null && periodo.isFazendaPublica()) {
        const inicial = periodo.getDataInicial();
        if (inicial !== null) {
          this.periodoDeJurosFazendaPublica = new Periodo(inicial, null);
          while (periodo !== null && periodo.isFazendaPublica()) {
            const fim: Date | null = periodo.getDataFinal();
            if (fim !== null) this.periodoDeJurosFazendaPublica.setFinal(fim);
            periodo = periodo.getProximoPeriodo();
          }
        }
      }
    }
    this.periodoDeJurosCalculado = true;
    return this;
  }

  /**
   * calcularTaxaDeJuros (Java linha 114).
   *
   * Delega para a tabela base do TabelaDeJuros chamando seu metodo via
   * `obterTaxaJuros(data)`. Aplica regras especificas:
   *  - Se `jurosDoAjuizamento` = TODAS, considera todas as competencias.
   *  - Se OCORRENCIAS_VENCIDAS, so aplica se data > dataInicialDeJuros.
   *  - Se `dataFimVencimentoOcorrencia` informado e data > fim, retorna null.
   *  - Multa/Honorario: NAO aplica `projetarData` (Java linha 117).
   */
  calcularTaxaDeJuros(
    data: Date,
    dataFimVencimentoOcorrencia: Date | null,
    jurosDoAjuizamento: JurosDoAjuizamentoEnum,
    _projetarData: boolean,
    _isMultaHonorario: boolean,
  ): Decimal | null {
    this.jurosDoAjuizamento = jurosDoAjuizamento;
    if (jurosDoAjuizamento === JurosDoAjuizamentoEnum.OCORRENCIAS_VENCIDAS) {
      if (this.dataInicialDeJuros && data.getTime() < this.dataInicialDeJuros.getTime()) {
        return null;
      }
    }
    if (dataFimVencimentoOcorrencia && data.getTime() > dataFimVencimentoOcorrencia.getTime()) {
      return null;
    }
    // Carrega a taxa do TabelaDeJurosDeInss (que ja tem TABELA_SELIC_MENSAL).
    // Por enquanto, sem instancia direta — usa fallback simples.
    return this.consultarTaxaSelic(data);
  }

  /** calcularTaxaDeJurosPagamento (Java linha 119). */
  calcularTaxaDeJurosPagamento(
    data: Date,
    dataFinal: Date,
    jurosDoAjuizamento: JurosDoAjuizamentoEnum,
    _projetarData: boolean,
    _isMultaHonorario: boolean,
  ): Decimal | null {
    this.jurosDoAjuizamento = jurosDoAjuizamento;
    // Pagamento usa intervalo data..dataFinal (acumulado).
    if (data.getTime() > dataFinal.getTime()) return null;
    return this.consultarTaxaSelic(data);
  }

  /**
   * Consulta a taxa SELIC mensal acumulada para uma data, usando a mesma
   * tabela do TabelaDeJurosDeInss (TABELA_SELIC_MENSAL). Acumula da data
   * informada ate a dataFinalParaCalculo (ou dataDeLiquidacao do calculo).
   */
  private consultarTaxaSelic(data: Date): Decimal | null {
    const fim = this.dataFinalParaCalculo ?? this.calculo.getDataDeLiquidacao() ?? new Date();
    let acumulado = 0;
    const cursor = new Date(Date.UTC(data.getFullYear(), data.getMonth(), 1));
    const limite = new Date(Date.UTC(fim.getFullYear(), fim.getMonth(), 1));
    while (cursor.getTime() <= limite.getTime()) {
      const entrada = TABELA_SELIC_MENSAL.find(
        e => e.ano === cursor.getFullYear() && e.mes === cursor.getMonth() + 1,
      );
      if (entrada) acumulado += entrada.taxa;
      cursor.setUTCMonth(cursor.getUTCMonth() + 1);
    }
    if (acumulado === 0) return null;
    return new Decimal(acumulado).div(100);
  }
}
